dask/dataframe/shuffle.py

```diff
@@ -963,57 +963,99 @@ def get_overlap(df, index):
 
         # Make sure that any data added from partition i-1 to `frames` is removed
         # from partition i-1.
-        dsk[(name, i - 1)] = (drop_overlap, dsk[(name, i - 1)], ddf.divisions[i])
+        dsk[(name, i - 1)] = (drop_overlap, dsk[(name, i - 1)], divisions[i])
 
         # We do not want to move "overlap" from the previous partition (i-1) into
         # this partition (i) if the data from this partition will need to be moved
         # to the next partition (i+1) anyway.  If we concatenate data too early,
         # we may lose rows (https://github.com/dask/dask/issues/6972).
-        if i == ddf.npartitions - 2 or ddf.divisions[i] != ddf.divisions[i + 1]:
-            frames.append((ddf._name, i))
+        if i == len(mins) - 2 or divisions[i] != divisions[i + 1]:
+            frames.append(dsk[(name, i)])
             dsk[(name, i)] = (methods.concat, frames)
             frames = []
 
     graph = HighLevelGraph.from_collections(name, dsk, dependencies=[ddf])
-    return new_dd_object(graph, name, ddf._meta, ddf.divisions)
+    return new_dd_object(graph, name, ddf._meta, divisions)
 
 
-def compute_and_set_divisions(df, **kwargs):
-    mins = df.index.map_partitions(M.min, meta=df.index)
-    maxes = df.index.map_partitions(M.max, meta=df.index)
-    mins, maxes = compute(mins, maxes, **kwargs)
+def _compute_partition_stats(column, allow_overlap=False, **kwargs) -> tuple:
+    """For a given column, compute the min, max, and len of each partition.
+
+    And make sure that the partitions are sorted relative to each other.
+    NOTE: this does not guarantee that every partition is internally sorted.
+    """
+    mins = column.map_partitions(M.min, meta=column)
+    maxes = column.map_partitions(M.max, meta=column)
+    lens = column.map_partitions(len, meta=column)
+    mins, maxes, lens = compute(mins, maxes, lens, **kwargs)
     mins = remove_nans(mins)
     maxes = remove_nans(maxes)
-
+    non_empty_mins = [m for m, l in zip(mins, lens) if l != 0]
+    non_empty_maxes = [m for m, l in zip(maxes, lens) if l != 0]
     if (
-        sorted(mins) != list(mins)
-        or sorted(maxes) != list(maxes)
-        or any(a > b for a, b in zip(mins, maxes))
+        sorted(non_empty_mins) != non_empty_mins
+        or sorted(non_empty_maxes) != non_empty_maxes
     ):
         raise ValueError(
-            "Partitions must be sorted ascending with the index", mins, maxes
+            f"Partitions must be sorted ascending by {column.name}",
+            f"In your dataset the mins of the partitions are: {mins}"
+            f"and the maxes are: {maxes}",
         )
+    if not allow_overlap and any(
+        a <= b for a, b in zip(non_empty_mins[1:], non_empty_maxes[:-1])
+    ):
+        raise ValueError(
+            "Partitions have overlapping values, so divisions are non-unique."
+            "Use `set_index(sorted=True)` with no `divisions` to allow dask to fix the overlap."
```

Comment:
Is no `divisions` necessary? Or would it be possible to print the divisions you should use in this error message, with duplicates removed, so that you can save having to do a compute pass over the data again?

If the divisions are sorted properly, but just have duplicates, then I think the only thing we can do right now is combine all the duplicates into one partition. Seems like the feeling in https://github.com/dask/dask/issues/8437 was that we can't handle duplicate divisions right now, so there's no other choice.

Reply:
I think you can either set `sorted=True` or you can set `divisions`. If you set both then we expect divisions to be exactly the right size for the current npartitions.

Reply:
Ah I see. You only get to go down the `fix_overlap` code path if you _don't_ pass in divisions. And you're thinking that doing that—even though it requires eagerly re-computing divisions again—will ultimately be more performant than a shuffle-based `set_index`, because it gets to use the `fix_overlap` process.

I'm not sure about that; I feel like the fact that `set_index(divisions=...)` would be fully lazy whereas `set_index(sorted=True)` would require re-computing the statistics you just computed might make the shuffle more worthwhile, even though you don't get to benefit from the fact that it's already sorted.

Reply:
Also, why make `allow_overlap=False` only warn?

Reply:
Yeah that makes sense about `set_index(divisions=...` being potentially better. When I was comparing them I was trying to minimize the number of tasks, but thinking it over this morning that clearly doesn't make sense - we are ok having more tasks if it means that we can be lazy.

> Also, why make `allow_overlap=False` only warn?

I wanted to be able to output the values even if they hit this scenario. Otherwise how will they know what to put into the `set_index(divisions=...)`? I'm fine to switch it back if you prefer to error. There is a test case that demonstrates a potentially confusing scenario: 

```python
    A = pd.DataFrame({"key": [1, 2, 3, 4, 4, 5, 6, 7], "value": list("abcd" * 2)})
    a = dd.from_pandas(A, npartitions=2)
    with pytest.warns(UserWarning, match="Partitions have overlapping values"):
        divisions = a.compute_current_divisions("key")
```

You'll notice that the "key" is sorted, it's just got an unfortunate duplicate right at the partition split so you end up with `[1, 2, 3, 4]` in the first partition and `[4, 5, 6, 7]` in the second. This counts as an overlap because if we just set `(0, 4, 7)` as the divisions we will drop rows in chained operations because dask won't know about the `4` in the first partition.

Reply:
Ahh I see. I think that's a good reason to keep it as a warning then. And then actually setting those divisions anywhere (or via `set_index`) should give you an error.